Haptics/app/src/main/java/com/example/android/haptics/samples/ui/views/HomeViewModel.kt

```diff
@@ -0,0 +1,239 @@
+    val vibrationEffectConstant: Int = -1
+)
+
+/**
+ * The category type, necessary since each haptic feedback effect calls the APIs differently.
+ */
+enum class HapticFeedbackCategoryType {
+    PREDEFINED_EFFECTS, HAPTIC_FEEDBACK_CONSTANTS, COMPOSITION_PRIMITIVES
+}
+
+/**
+ * ViewModel that handles the business logic of the home screen.
+ */
+class HomeViewModel(
+    private val vibrator: Vibrator,
+    val viewModelState: HomeViewModelState,
+    private val scaffoldState: ScaffoldState,
+    val scrollState: ScrollState
+) : ViewModel() {
+
+    /**
+     * On each button click on the home page, the composable calls this event, passing along all
+     * the necessary information to call the correct haptics API.
+     */
+    fun onHapticFeedbackButtonClicked(view: View, category: HapticFeedbackCategoryType, effect: Int) {
+        if (category == HapticFeedbackCategoryType.PREDEFINED_EFFECTS) {
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+                // https://developer.android.com/reference/android/os/VibrationEffect#createPredefined(int)
+                vibrator.vibrate(VibrationEffect.createPredefined(effect))
+            }
+        } else if (category == HapticFeedbackCategoryType.HAPTIC_FEEDBACK_CONSTANTS) {
+            // https://developer.android.com/reference/android/view/View#performHapticFeedback(int)
+            view.performHapticFeedback(effect)
+        } else if (category == HapticFeedbackCategoryType.COMPOSITION_PRIMITIVES) {
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
+                // https://developer.android.com/reference/android/os/VibrationEffect.Composition
+                vibrator.vibrate(
+                    VibrationEffect.startComposition()
+                        .addPrimitive(effect)
+                        .compose()
+                )
+            }
+        } else {
+            throw RuntimeException("No corresponding action for HapticFeedbackCategoryType.$category.")
+        }
+    }
+
+    fun onSnackbarMessage(message: String) {
+        viewModelScope.launch {
+            scaffoldState.snackbarHostState.showSnackbar(message)
+        }
+    }
+
+    /**
+     * Factory for HomeViewModel that can provide vibrator dependency and construct
+     * HomeViewModelState using Application context.
+     */
+    companion object {
+        fun provideFactory(
+            application: Application,
+            scaffoldState: ScaffoldState,
+            scrollState: ScrollState,
+        ): ViewModelProvider.Factory = object : ViewModelProvider.Factory {
+            @Suppress("UNCHECKED_CAST")
+            override fun <T : ViewModel> create(modelClass: Class<T>): T {
+                val isAndroidQOrLater = Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q // API Level 29.
+                val isAndroidROrLater = Build.VERSION.SDK_INT >= Build.VERSION_CODES.R // API Level 30.
```

Comment:
Done, added them to the companion object so they can be reused. I actually couldn't use them as constants otherwise compiler would complain everywhere where I checked.